base/array.jl

```diff
@@ -834,6 +834,43 @@ function setindex!(A::Array{T}, X::Array{T}, c::Colon) where T
     return A
 end
 
+"""
+    setindex(collection, value, key...)
+
+Create a new collection with the element/elements of the location specified by `key`
+replaced with `value`(s).
+
+# Implementation
+
+`setindex(collection, value, key...)` must have the property such that
+
+```julia
+y1 = setindex(x, value, key...)
+
+y2 = deepcopy(x)
+setindex!(y2, value, key...)
+
+@assert isequal′(y1, y2)
+```
+
+with a suitable definition of `isequal′` (e.g., elements are approximately but possibly
+not exactly equal), _if_ `setindex!` supports given arguments.
```

Comment:
The not exactly equal part here is just because we have, e.g., `convert(Float64, Int(maxintfloat()) + 1) != Int(maxintfloat()) + 1`, right?  Maybe it'd be a bit clearer to directly say that it's equal after conversion?  I think I'd rather focus on the `copy` side of things than the equality side of things:

That is, something more like:

```julia
y1 = setindex(x, value, key...)
y2 = copy′(x)
y2[key...] = value

@assert convert.(eltype(y2), y1) == y2
```

with a suitable definition of `copy′` that ensures `value` can be assigned into the collection.  This is where the focus should be.  Getting the `@assert` exactly right (which it's not right now because broadcast) is slightly less important than getting the widening right, I think.

That way we don't have to bend over backwards with awkward approximately-but-maybe-not-exact sorts of phrasings.  It's just a (possible `convert` plus) `copy` plus `setindex!`.

Reply:
This is a great idea! Yes, it was a very awkward part of my version of the specification. Your version is much more elegant and precise.